src/github.com/matrix-org/dendrite/syncapi/sync/notifier.go

```diff
@@ -15,53 +15,217 @@
+	"github.com/matrix-org/dendrite/syncapi/storage"
 	"github.com/matrix-org/dendrite/syncapi/types"
 	"github.com/matrix-org/gomatrixserverlib"
 )
 
-// Notifier will wake up sleeping requests in the request pool when there
-// is some new data. It does not tell requests what that data is, only the
-// stream position which they can use to get at it.
+// Notifier will wake up sleeping requests when there is some new data.
+// It does not tell requests what that data is, only the stream position which
+// they can use to get at it. This is done to prevent races whereby we tell the caller
+// the event, but the token has already advanced by the time they fetch it, resulting
+// in missed events.
 type Notifier struct {
-	// The latest sync stream position: guarded by 'cond'.
-	currPos types.StreamPosition
-	// A condition variable to notify all waiting goroutines of a new sync stream position
-	cond *sync.Cond
+	// The latest sync stream position: guarded by 'currPosMutex' which is RW to allow
+	// for concurrent reads on /sync requests
+	currPos      types.StreamPosition
+	currPosMutex *sync.RWMutex
+	// A map of RoomID => Set<UserID> : Map access is guarded by roomIDToJoinedUsersMutex.
+	roomIDToJoinedUsers      map[string]set
+	roomIDToJoinedUsersMutex *sync.Mutex
+	// A map of user_id => UserStream which can be used to wake a given user's /sync request.
+	// Map access is guarded by userStreamsMutex.
+	userStreams      map[string]*UserStream
+	userStreamsMutex *sync.Mutex
 }
 
 // NewNotifier creates a new notifier set to the given stream position.
+// In order for this to be of any use, the Notifier needs to be told all rooms and
+// the joined users within each of them by calling Notifier.LoadFromDatabase().
 func NewNotifier(pos types.StreamPosition) *Notifier {
 	return &Notifier{
-		pos,
-		sync.NewCond(&sync.Mutex{}),
+		currPos:                  pos,
+		currPosMutex:             &sync.RWMutex{},
+		roomIDToJoinedUsers:      make(map[string]set),
+		roomIDToJoinedUsersMutex: &sync.Mutex{},
+		userStreams:              make(map[string]*UserStream),
+		userStreamsMutex:         &sync.Mutex{},
 	}
 }
 
 // OnNewEvent is called when a new event is received from the room server. Must only be
 // called from a single goroutine, to avoid races between updates which could set the
 // current position in the stream incorrectly.
 func (n *Notifier) OnNewEvent(ev *gomatrixserverlib.Event, pos types.StreamPosition) {
-	// update the current position in a guard and then notify all /sync streams
-	n.cond.L.Lock()
+	// update the current position in a guard and then notify relevant /sync streams.
+	// This needs to be done PRIOR to waking up users as they will read this value.
+	n.currPosMutex.Lock()
 	n.currPos = pos
-	n.cond.L.Unlock()
+	n.currPosMutex.Unlock()
 
-	n.cond.Broadcast() // notify ALL waiting goroutines
+	// Map this event's room_id to a list of joined users, and wake them up.
+	userIDs := n.joinedUsers(ev.RoomID())
```

Comment:
No, because those functions are only called by the same goroutine that call this function (it all happens on the consumer goroutine).